### PR TITLE
Add FP-DDM thermal example and numerical elasticity baseline

### PR DESCRIPTION
**Pull request by loliverhennigh**
## Summary

FP-DDM (Foundation-model-based Physics-guided adaptation for the Domain Decomposition Method) is a scalable framework for large-domain multi-physics analysis.

This example demonstrates the PhysicsNeMo integration of the FP-DDM domain-decomposition work proposed in DAC 2026 [1] by the Samsung CSE team and UNIST.

- Add a compact FP-DDM thermal example with synthetic material fields, overlapping subdomains, Schwarz iteration, three interface-update options, a matrix-free finite-volume reference solver, and an optional PhysicsNeMo FNO local solver.
- Add a numerical two-dimensional plane-stress baseline that compares a 2 x 2 overlapping DDM solve with a monolithic solve using the same matrix-free elasticity operator.
- Use native PhysicsNeMo components for the FNO, `Module`, Hydra configuration, `DistributedManager`, `LaunchLogger`, and checkpoint save/load.
- Consolidate the original 55-file draft into 19 example files, plus one catalog entry and two focused test modules. Remove dynamic registries, compatibility aliases, duplicate package layers, and the misleading Lightning naming.
- Preserve the original one-stage physics-informed training and gradient-based FP-DDM workflow while fixing coordinate orientation, heat-source residuals, boundary inputs, test-time adaptation state and batch normalization, interface normalization, batched solver convergence, metric ordering and persistence, and heat-flux diagnostics.

## Scope

The default thermal workload retains zero heat source, a 32 x 32 local grid, the original FNO architecture, the parallel gradient-based interface update, and optional test-time adaptation disabled by default. Full-domain reference solves are opt-in and bounded to small decompositions. The example remains single-process for Schwarz inference and keeps global fields for diagnostics; it does not claim billion-cell production scaling.

The elasticity path is a numerical baseline for the FP-DDM workflow. It does not train or use an elasticity FNO, so it is not a neural FP-DDM result.

## Validation

- `pytest -q test/examples/test_fp_ddm.py test/examples/test_fp_ddm_elasticity.py` (17 passed)
- Ruff check and format
- Docstring coverage, Markdown, license, added-file, and import-contract checks
- Regression coverage for batch-size-invariant test-time adaptation, incremental metric writes with clean rerun behavior, and explicit reference-solve bounds
- Default 32 x 32 elasticity comparison: converged in 64 DDM iterations; displacement relative error 2.781e-6 and stress relative error 6.591e-6 against the monolithic solve
- One-epoch training/checkpoint/reload smoke on CPU
- Checkpoint-backed FNO run with one test-time adaptation step
- Full visualization run producing PNG, NumPy, convergence, heat-flux, and MP4 artifacts
- A100 80 GB training baseline: 500,000 configured samples, batch size 256, and the selected epoch-6 checkpoint; approximately 3 minutes to epoch 6
- Ten independently generated 3 x 3 layouts: mean iteration-50 NRMAE 0.0361, worst NRMAE 0.0386, and mean R2 0.9495
- Five independently generated 5 x 5 layouts: mean iteration-50 NRMAE 0.0540, worst NRMAE 0.0607, and mean R2 0.9157
- 3 x 3 finite-volume rollout: iteration-50 NRMAE 0.00591 and R2 0.9988

## Known limitation

Neural checkpoint quality is not monotonic in the isolated local validation loss. The lower-validation-loss epoch-9 checkpoint reached mean NRMAE 0.3091 at iteration 50, while epoch 6 with the existing parallel handler at learning rate 0.5 remained effectively monotonic and reached mean NRMAE 0.0361. A seed-10 stress run eventually drifted from NRMAE 0.0360 at iteration 54 to 0.0616 at iteration 200 even as interface disagreement approached zero. The README reports these results and recommends selecting checkpoints and stopping settings on held-out decomposed problems.

### CODE REVIEW

File: examples/tcad/fp_ddm/fpddm/schwarz.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                        prediction,
+                        normalized,
+                        source_scale=model.nondimensional_source_scale,
+                    )
+                    total = (
+                        self.pde_weight * pde_loss
+                        + self.boundary_weight * boundary_loss
+                    )
+                    total.backward()
+                optimizer.step()
```
[the remark this answers is not in the document]
Comment (loliverhennigh) [reply examples/tcad/fp_ddm/fpddm/schwarz.py:92]: Fixed in d4a633d4. Each mini-batch loss is now weighted by its fraction of the adapted subdomains before backward(), and the reported TTT loss is sample-weighted as well. The regression test compares one full batch with split batches using SGD so an objective-scaling difference is observable.

File: examples/tcad/fp_ddm/fpddm/observers.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        for getter in self.metric_sources:
+            name, value = getter()
+            row[name] = float(value)
+        self.rows.append(row)
+        self.csv_path.parent.mkdir(parents=True, exist_ok=True)
+        with self.csv_path.open("w", newline="", encoding="utf-8") as output:
+            writer = csv.DictWriter(output, fieldnames=list(self.rows[-1]))
+            writer.writeheader()
+            writer.writerows(self.rows)
+        return row
```
[the remark this answers is not in the document]
Comment (loliverhennigh) [reply examples/tcad/fp_ddm/fpddm/observers.py:238]: Fixed in d4a633d4. MetricsLogger now truncates and writes the header on the first iteration, then appends exactly one row per step while retaining the in-memory history used by RunResult. The regression test verifies w/a modes and clean rerun behavior against a stale CSV.

**Comment by loliverhennigh** [comment]
Follow-up on the remaining automated review notes:

- Full-domain reference solves are now explicitly opt-in (`run.ground_truth=false` by default). Requests above the existing 26-subdomain bound emit a `RuntimeWarning` and continue without reference metrics; the README documents the bound and the small validation commands opt in explicitly.
- I retained `ThermalFEMSolver` and the `fem` configuration key to preserve the original contribution and command-line compatibility. The class docstring, config, and README identify the implementation as a matrix-free finite-volume solver. Renaming that public surface would add migration churn without changing the numerical implementation.

The focused suite now has 11 passing tests, including regressions for these cleanup changes.

**Comment by loliverhennigh** [comment]
CI status on the rebased head (`d4a633d4`): NVIDIA `Run pre-commit hooks` and `Test with uv` both pass; the pip matrix is intentionally skipped. The separate pre-commit.ci integration fails before running import-linter because its environment has no `lint-imports` executable. This is repository-wide rather than introduced here: merged PR #1787 has the same pre-commit.ci failure while the NVIDIA jobs pass.

**Comment by with2say** [comment]
Hi @loliverhennigh 
Thank you for the careful consolidation. The example faithfully preserves the FP-DDM workflow from our paper (one-stage physics-informed training + gradient-based interface updates).

We ran the branch end-to-end and confirmed that the PhysicsNeMo port reproduces the expected FP-DDM behavior and accuracy.

On credit alignment, we'd suggest adding the following to the README intro and to the top of the summary:

1. **Credit** — This example demonstrates the PhysicsNeMo integration of the FP-DDM domain-decomposition work proposed in DAC 2026 [1] by the Samsung CSE team and UNIST.

2. **Meaning of FP-DDM** — FP-DDM (Foundation-model-based Physics-guided adaptation for the Domain Decomposition Method), a scalable framework for large-domain multi-physics analysis.

3. **Reference** — a Reference section in the example README with the BibTeX below. Note the DOI is not active yet; it will resolve once the DAC '26 proceedings are published.

   ```bibtex
   @inproceedings{park2026fpddm,
     author    = {Park, Min-Chul and Lee, Ji-Hye and Park, Hong-Hyun and Huh, In and Lee, Sungyeop and Jang, Hyunjae and Hong, Giyong and Lee, Seokki and Jeong, Changwook and Kim, Young-Gu and Kim, Dae Sin},
     title     = {Neural Domain Decomposition for Scalable Multi-Physics: Chip-Scale Thermal-Stress Analysis},
     booktitle = {Proceedings of the 63rd ACM/IEEE Design Automation Conference (DAC '26)},
     year      = {2026},
     doi       = {10.1145/3770743.3803918}
   }
   ```

Once the credit and reference are reflected, we'd like to do a quick internal review of the final wording on our side before the merge. Happy to review anything further.

**Comment by Kihang-Youn** [comment]
Hi @ram-cherukuri  @ktangsali, Could you please take a look when you have a chance? Thanks!

File: examples/tcad/fp_ddm/fpddm/model.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    grad_x = torch.gradient(temperature, dim=3)[0]
+    grad_y = torch.gradient(temperature, dim=2)[0]
+    flux_x = conductivity * grad_x
+    flux_y = conductivity * grad_y
+    div_x = torch.gradient(flux_x, dim=3)[0]
+    div_y = torch.gradient(flux_y, dim=2)[0]
+    return div_x + div_y + source_scale * heat_source
+
+
+def thermal_losses(
```
Comment (ktangsali) [inline examples/tcad/fp_ddm/fpddm/model.py:61]: Could we have used the `PhysicsInformer` here? Or atleast the gradient functionals?

File: examples/tcad/fp_ddm/fpddm/solvers.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        *,
+        max_iter: int = 1000,
+        tolerance: float = 1.0e-12,
+        device: str | torch.device | None = None,
+    ) -> None:
+        """Configure local solve batching and convergence."""
+
+        super().__init__(device=device)
+        self.batch_size = batch_size
+        self.solver = Heat2DSolver(
```
Comment (ktangsali) [inline examples/tcad/fp_ddm/fpddm/solvers.py:112]: The finite-volume reference solver appears to use its default dx=dy=1. The model inputs describe a normalized [0, 1]² patch, so the reference solver and neural residual should derive their spacing from the same coordinate convention. Otherwise, they solve differently scaled equations when the heat source is nonzero.

File: examples/tcad/fp_ddm/fpddm/training.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            if validation_metrics["loss"] <= best_loss:
+                best_loss = validation_metrics["loss"]
+                save_checkpoint(
+                    best_dir,
+                    models=core,
+                    epoch=epoch,
+                    metadata={"best_validation_loss": best_loss},
+                )
+
+        if deadline is not None and time.monotonic() >= deadline:
```
Comment (ktangsali) [inline examples/tcad/fp_ddm/fpddm/training.py:245]: Please check if synchronization of this stop decision is needed across ranks

**Review by ktangsali — APPROVED** [review]
Looks good! Left a few comments

File: examples/tcad/fp_ddm/fpddm/solvers.py (hunk 4 above)
Comment (loliverhennigh) [reply examples/tcad/fp_ddm/fpddm/solvers.py:114]: Fixed in c28b9bcf. Good catch—the zero-source default hid this. The neural residual now derives spacing from its normalized coordinate channels, while the finite-volume solver uses the same base patch spacing for local and full-domain solves and scales it for the coarse downsampled solve. A nonzero-source manufactured solution (-laplacian(T)=4 on the unit square) matches with maximum absolute error 2.1e-7, and a nonzero-source 2x2 end-to-end FP-DDM smoke run passes.

File: examples/tcad/fp_ddm/fpddm/training.py (hunk 5 above)
Comment (loliverhennigh) [reply examples/tcad/fp_ddm/fpddm/training.py]: Fixed in c28b9bcf. The deadline decision is now reduced with all_reduce(MAX) after the rank-0 checkpoint block, so every rank exits the epoch loop together if any rank reaches the time budget. I also exercised the helper with a real two-rank Gloo process group.

File: examples/tcad/fp_ddm/fpddm/model.py (hunk 3 above)
Comment (loliverhennigh) [reply examples/tcad/fp_ddm/fpddm/model.py:63]: I looked into this. PhysicsInformer is conceptually applicable, but its structured finite-difference path currently operates on one unbatched field, and the native grid derivative functionals are likewise unbatched and periodic. This loss runs on batched Dirichlet patches and intentionally evaluates the flux form div(k grad(T)), so switching would not be behavior-preserving without a larger refactor. I retained the vectorized torch.gradient implementation; c28b9bcf now derives its spacing explicitly from the normalized coordinate channels.

**Comment by loliverhennigh** [comment]
@greptileai review the latest changes

**Comment by loliverhennigh** [comment]
@ktangsali All three review comments are addressed, the threads are resolved, and the PR is updated to current main at 2fa3acc8. The public checks are green, but the merge queue is still waiting for the trusted copy-pr-bot run. Could you please run /ok to test for this head when you have a chance?

**Comment by loliverhennigh** [comment]
/ok to test [2fa3acc8](https://github.com/NVIDIA/physicsnemo/pull/1788/commits/2fa3acc8c79327a5b8708fa16cc383f94b135288)
